corehq/apps/geospatial/models.py

```diff
@@ -1 +1,130 @@
-# Create your models here.
+from math import radians, cos, sin, asin, sqrt
+from shapely.geometry import Point
+import geopandas as gpd
+
+
+class GeoObject:
+    """
+    `lat` and `lon` are assumed to be in a EPSG:4326 map projection.
+    """
+    def __init__(self, id, lon, lat):
+        if lon < -180 or lon > 180:
+            raise ValueError("Invalid lon value. Must be [-180, 180]")
+        if lat < -90 or lat > 90:
+            raise ValueError("Invalid lat value. Must be [-90, 90]")
+
+        self.id = id
+        self.lat = lat
+        self.lon = lon
+
+    def get_point(self):
+        return Point(self.lon, self.lat)
+
+    def get_info(self):
+        return {
+            'id': self.id,
+            'geometry': self.get_point()
+        }
+
+
+class Objective(GeoObject):
+    is_assigned = False
+
+    def get_info(self):
+        info = super().get_info()
+        info['is_assigned'] = self.is_assigned
+        return info
+
+
+class ObjectiveAllocator:
+    """
+    :param users: Should be a list of objects of class/subclass `GeoObject` with id of `user_id`
+    :param objectives: Should be a list of objects of class/subclass `GeoObject` with id of `case_id`
+    :param max_distance: The maximum distance allowed for an objective to be assignable to a user (unit is km)
+    :param max_assignable: The maximum number of objectives that can be assigned to a user
+    """
+    def __init__(self, users, objectives, max_distance, max_assignable):
+        if max_distance < 0 or max_assignable < 0:
+            raise ValueError("Maximum distance and assignable count must be positive numbers")
+
+        self.users = users
+        self.objectives = objectives
+        self.max_distance = max_distance
+        self.max_assignable = max_assignable
+
+    def _create_dataframe(self, item_list):
+        return gpd.GeoDataFrame([item.get_info() for item in item_list]).set_index('id')
+
+    def _get_distance_between_points(self, start_point, end_point):
```

Comment:
@zandre-eng 
It looks like geopandas [supports distance calculation](https://stackoverflow.com/a/63725180) between any two points. Any reason why we're not using that?

Reply:
It looks like we have to create a dataframe containing each set of points we want to get the distance to, so not sure if that might create unnecessary overhead as opposed to manually calculating it as you're doing here.